### PR TITLE
fix: Remove credentials pinned to Vault, LoanBroker, and AMM pseudo-accounts

### PR DESCRIPTION
**Pull request by tyalymov**
## High Level Overview of Change

A credential issued to a Vault, LoanBroker, or AMM pseudo-account can never be accepted or deleted by it, because pseudo-accounts can't sign transactions. The credential stays pinned in the pseudo-account's owner directory and blocks deletion of the owning object, locking up its owner reserve. Creating such a credential is now rejected on its own amendment; this change adds the cleanup for pins that predate that rejection.

### Context of Change

`VaultDelete` and `LoanBrokerDelete` require the pseudo-account's owner directory to be empty and fail with `tecHAS_OBLIGATIONS` when a stale credential is still linked there. AMM fails earlier and harder: a sole-LP withdrawal walks the AMM account's owner directory and returns `tecINTERNAL` on the unexpected credential, so a pinned AMM can't even be fully withdrawn.

Behind a new amendment, `VaultDelete` and `LoanBrokerDelete` delete any credentials pinned to the pseudo-account before removing it. The cleanup is bounded per transaction; a large pin resumes across multiple transactions via `tecINCOMPLETE`. `deleteAMMTrustLines` removes AMM's pinned credentials the same way, and the sole-LP withdrawal check now ignores credential entries so a pinned AMM can still be withdrawn.

### API Impact

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [x] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

## Test Plan

Added tests to the Vault, LoanBroker, and AMM suites: each pins a credential to the pseudo-account, confirms deletion is blocked, then activates the amendment and confirms the object deletes cleanly with the credential removed. A further test pins more credentials than the per-transaction bound and confirms cleanup finishes across two transactions.

### CODE REVIEW

File: src/libxrpl/tx/transactors/vault/VaultDelete.cpp [hunk 1]
```diff
@@ -158,6 +159,16 @@ VaultDelete::doApply()
 
     view().erase(mpt);
 
+    // Remove any credentials pinned to the vault pseudo-account. They can
+    // otherwise keep its owner directory alive and block deletion with
+    // tecHAS_OBLIGATIONS. See fixCredentialPseudoAccount.
+    if (view().rules().enabled(fixCredentialPseudoAccount))
```
[the remark this answers is not in the document]
Comment (tyalymov) [reply src/libxrpl/tx/transactors/vault/VaultDelete.cpp]: The bot is right that AMM was missed: part 1 rejects AMM pseudo-account subjects, so part 2 has to cover AMM deletion too, otherwise a credential pinned before the amendment would strand the AMM.

Where the bot is off is the mechanism. The failure doesn't start in `deleteAMMTrustLines`. With a credential on the AMM pseudo-account, a full withdrawal fails with `tecINTERNAL` earlier, in `isOnlyLiquidityProvider` (called from `verifyAndAdjustLPTokenBalance` on the sole-LP withdrawal path). It walks the AMM account's owner directory and returns `tecINTERNAL` on any entry that isn't `ltAMM`/`ltMPTOKEN`/`ltRIPPLE_STATE`, so the withdrawal never reaches `deleteAMMAccount`. The cleanup the bot suggested, added only there, wouldn't have helped.

What I did, both gated behind `fixCredentialPseudoAccount`:
- `isOnlyLiquidityProvider` now skips `ltCREDENTIAL` entries, so the withdrawal math tolerates a pre-amendment pin.
- `deleteAMMAccount` removes the pinned credentials before deleting the trust lines, reusing the same `deletePseudoAccountCredentials` helper as Vault/LoanBroker.

Added an AMM test that reproduces the strand with the amendment off, then confirms the new pin is rejected and the AMM deletes cleanly with it on.

@Tapanito flagging the `isOnlyLiquidityProvider` change specifically, since it sits in the withdrawal path and could use a second pair of eyes.

File: src/libxrpl/ledger/helpers/CredentialHelpers.cpp [hunk 2]
```diff
@@ -123,6 +124,28 @@ deleteSLE(ApplyView& view, SLE::ref sleCredential, beast::Journal j)
     return tesSUCCESS;
 }
 
+TER
+deletePseudoAccountCredentials(ApplyView& view, AccountID const& pseudoAcct, beast::Journal j)
+{
+    // Pseudo-accounts can't sign, so any credential linked into their owner
```
Comment (Tapanito) [inline src/libxrpl/ledger/helpers/CredentialHelpers.cpp]: While the method name suggests thta credentials are deleted for pseudoAccounts, in reality, any accountID can be passed. 

It might be a good idea to check that the account passed is a pseudo-account.

File: include/xrpl/ledger/helpers/CredentialHelpers.h [hunk 3]
```diff
@@ -32,6 +32,13 @@ checkExpired(SLE const& sleCredential, NetClock::time_point const& closed);
 [[nodiscard]] TER
 deleteSLE(ApplyView& view, SLE::ref sleCredential, beast::Journal j);
 
+// Remove every credential linked into a pseudo-account's owner directory.
+// Pseudo-accounts can't sign, so any such credential names the pseudo-account
+// as subject and can never be accepted or deleted by it. Non-credential objects
+// are left in place; the caller decides whether they block deletion.
```
Comment (Tapanito) [inline include/xrpl/ledger/helpers/CredentialHelpers.h]: nit: Please use doxygen style comment.

Comment (Tapanito) [inline include/xrpl/ledger/helpers/CredentialHelpers.h]: Since after this PR `CredentialCreate` would not allow creating credentials for a pseudo-account, this function, at least for now, will only be useful, until the amendment that introduced it, is retired. It would be great to document why this method is needed (in case there are any pseudo-accounts already locked up).

File: include/xrpl/protocol/detail/features.macro [hunk 4]
```diff
@@ -15,6 +15,7 @@
 // Add new amendments to the top of this list.
 // Keep it sorted in reverse chronological order.
 
+XRPL_FIX    (CredentialPseudoAccount,     Supported::Yes, VoteBehavior::DefaultNo)
```
Comment (Tapanito) [inline include/xrpl/protocol/detail/features.macro]: Each amendment requires to be voted in. Since `Cleanup3_4_0` is not yet released, let's add the fix under it, rather than a new amendment.

File: src/libxrpl/ledger/helpers/CredentialHelpers.cpp [hunk 5]
```diff
@@ -123,6 +124,28 @@ deleteSLE(ApplyView& view, SLE::ref sleCredential, beast::Journal j)
...
+    // directory names the pseudo-account as subject and can never be accepted
+    // or deleted by it. Remove those credentials so the pseudo-account (and its
+    // owning object) can be deleted. Other objects are left untouched: the
+    // caller's own checks decide whether the remaining directory blocks
+    // deletion.
+    return cleanupOnAccountDelete(
```
[the remark this answers is not in the document]
Comment (Tapanito) [reply src/libxrpl/ledger/helpers/CredentialHelpers.cpp:148]: Good catch, clanker!

@tyalymov  once maxNodesToDelete is added, let's add a test analogous to the existing kMaxDeletableAmmTrustLines overflow tests (AMM_test.cpp:5214, :7192) that pins more than the cap's worth of credentials to a pseudo-account pre-amendment and verifies tecINCOMPLETE + eventual multi-transaction cleanup.
- Codecov flags one uncovered line each in AMMHelpers.cpp, LoanBrokerDelete.cpp, and VaultDelete.cpp — all the !isTesSuccess(ter) early-return branches after calling deletePseudoAccountCredentials. Once the function can return tecINCOMPLETE (per the Must Fix above), these branches become reachable and should be covered by the new test.

**Review by Tapanito — CHANGES_REQUESTED** [review]
Nicely done, some minor comments.

File: src/libxrpl/ledger/helpers/CredentialHelpers.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (tyalymov) [reply src/libxrpl/ledger/helpers/CredentialHelpers.cpp:148]: Thanks for catching this. I started implementing the bound and hit a snag I'd like to consult you on.

The plan was to delete up to N credentials per transaction and return `tecINCOMPLETE` so a follow-up transaction resumes the cleanup. That works for LoanBroker and AMM. For Vault it collides with `ValidMPTIssuances`: `VaultDelete` carries the `DestroyMptIssuance` privilege, and the invariant requires any applied `VaultDelete` to remove exactly one MPT issuance in the same transaction. An incremental `VaultDelete` that clears a batch of credentials and returns `tecINCOMPLETE` hasn't removed the share issuance yet, so it trips `tecINVARIANT_FAILED`.

Two directions I came up with:

1. Relax the invariant to require the issuance removal only on a `tesSUCCESS` `VaultDelete`, not on an in-progress `tecINCOMPLETE` one. It's a small change and a no-op before the amendment (`VaultDelete` can't return `tecINCOMPLETE` without it), but it does touch a consensus invariant.
2. Refuse the delete when the pseudo-account holds more than the cap, like `AccountDelete`'s `tefTOO_BIG`, with no partial progress. Simpler, but a Vault pinned with more than the cap could then never be deleted, which doesn't fully cure the pre-amendment case.

LoanBroker and AMM I can bound incrementally regardless; it's the Vault path I'm unsure about. Any thoughts on that?

[the remark this answers is not in the document]
Comment (Tapanito) [reply src/libxrpl/ledger/helpers/CredentialHelpers.cpp:148]: The first direction sounds right, in the transactor we first attempt to delete the credentials. If we fail to do so because there are too many of them, then the user must submit the transaction again to delete the rest of the credentials and the account.

File: src/libxrpl/tx/transactors/vault/VaultDelete.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tyalymov) [reply src/libxrpl/tx/transactors/vault/VaultDelete.cpp]: Fixed — AMM is covered now: `isOnlyLiquidityProvider` skips credential entries, and `deleteAMMTrustLines` removes them within its existing bounded walk. (The actual failure point turned out to be `isOnlyLiquidityProvider` during withdrawal rather than `deleteAMMTrustLines`, as noted above.)

File: src/libxrpl/ledger/helpers/CredentialHelpers.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (tyalymov) [reply src/libxrpl/ledger/helpers/CredentialHelpers.cpp:148]: Fixed

File: src/libxrpl/ledger/helpers/CredentialHelpers.cpp [hunk 6]
```diff
@@ -123,6 +124,28 @@ deleteSLE(ApplyView& view, SLE::ref sleCredential, beast::Journal j)
     return tesSUCCESS;
 }
 
+TER
```
[the remark this answers is not in the document]
Comment (tyalymov) [reply src/libxrpl/ledger/helpers/CredentialHelpers.cpp:132]: Fixed

File: src/libxrpl/ledger/helpers/CredentialHelpers.cpp (hunk 2 above)
Comment (tyalymov) [reply src/libxrpl/ledger/helpers/CredentialHelpers.cpp]: Fixed

File: include/xrpl/ledger/helpers/CredentialHelpers.h (hunk 3 above)
Comment (tyalymov) [reply include/xrpl/ledger/helpers/CredentialHelpers.h]: Fixed

Comment (tyalymov) [reply include/xrpl/ledger/helpers/CredentialHelpers.h]: Fixed

File: include/xrpl/protocol/detail/features.macro (hunk 4 above)
Comment (tyalymov) [reply include/xrpl/protocol/detail/features.macro]: Fixed

File: src/libxrpl/tx/transactors/credentials/CredentialCreate.cpp [hunk 7]
```diff
@@ -90,6 +90,16 @@ CredentialCreate::preclaim(PreclaimContext const& ctx)
         return tecNO_TARGET;
     }
 
+    // A pseudo-account (Vault, LoanBroker, AMM) can't sign, so it can never
+    // accept or delete a credential issued to it. Such a credential would stay
+    // pinned in the pseudo-account's owner directory forever and block deletion
+    // of the owning object (tecHAS_OBLIGATIONS). Reject it up front.
+    if (ctx.view.rules().enabled(fixCleanup3_4_0) && isPseudoAccount(ctx.view, subject))
```
Comment (mvadari) [inline src/libxrpl/tx/transactors/credentials/CredentialCreate.cpp]: FYI this is included in 3.3 (was a later PR so didn't make it into the public rc), so perhaps wait to merge this PR until we merge 3.3 into develop

**Review by mvadari — DISMISSED** [review]

File: src/libxrpl/tx/transactors/lending/LoanBrokerDelete.cpp [hunk 8]
```diff
@@ -141,6 +143,19 @@ LoanBrokerDelete::doApply()
...
+    // else. They would otherwise keep its owner directory alive and block
+    // deletion with tecHAS_OBLIGATIONS. Doing it first means a bounded,
+    // tecINCOMPLETE cleanup can be resumed by a later transaction without having
+    // already torn down the broker. See fixCleanup3_4_0.
+    if (view().rules().enabled(fixCleanup3_4_0))
+    {
+        if (auto const ter = credentials::deletePseudoAccountCredentials(
+                view(), brokerPseudoID, kMaxDeletablePseudoAccountCredentials, j_);
+            !isTesSuccess(ter))
+            return ter;
```
Comment (a1q123456) [inline src/libxrpl/tx/transactors/lending/LoanBrokerDelete.cpp:155]: I think you can either create a corrupt dir to forge an error, or just mark this line as // LCOV_EXCL_LINE

**Review by a1q123456 — CHANGES_REQUESTED** [review]
Looks good except for a small issue.

File: src/libxrpl/tx/transactors/credentials/CredentialCreate.cpp (hunk 7 above)
Comment (tyalymov) [reply src/libxrpl/tx/transactors/credentials/CredentialCreate.cpp]: Thanks for the flag. 3.3 is in develop now, and this PR no longer overlaps with it: I dropped the create-time rejection from here entirely and kept only the cleanup for credentials pinned before that rejection existed.

File: src/libxrpl/tx/transactors/lending/LoanBrokerDelete.cpp (hunk 8 above)
Comment (tyalymov) [reply src/libxrpl/tx/transactors/lending/LoanBrokerDelete.cpp:155]: Added a test that pins more credentials than the per-transaction bound, so this line is exercised for real (tecINCOMPLETE) rather than needing LCOV_EXCL_LINE.

**Review by Tapanito — APPROVED** [review]

**Review by a1q123456 — APPROVED** [review]
lgtm

File: include/xrpl/protocol/Protocol.h [hunk 9]
```diff
@@ -396,6 +396,12 @@ using TxID = uint256;
  */
 constexpr std::uint16_t kMaxDeletableAmmTrustLines = 512;
 
+/**
+ * The maximum number of credentials to delete from a pseudo-account's owner
+ * directory in a single transaction.
+ */
```
[the remark this answers is not in the document]
Comment (tyalymov) [reply include/xrpl/protocol/Protocol.h:406]: Fixed

File: include/xrpl/ledger/helpers/CredentialHelpers.h [hunk 10]
```diff
@@ -33,6 +34,29 @@ checkExpired(SLE const& sleCredential, NetClock::time_point const& closed);
...
 deleteSLE(ApplyView& view, SLE::ref sleCredential, beast::Journal j);
 
+/**
+ * @brief Remove credentials pinned to a pseudo-account's owner directory.
+ *
+ * Cleans up credentials that were linked to a pseudo-account (Vault, LoanBroker,
+ * AMM), which such an account can neither accept nor delete. Only credentials
+ * are removed, at most @p maxNodesToDelete of them; on reaching that bound the
+ * result is `tecINCOMPLETE` and the caller must propagate it so a later
+ * transaction resumes.
```
[the remark this answers is not in the document]
Comment (tyalymov) [reply include/xrpl/ledger/helpers/CredentialHelpers.h]: Fixed

File: src/libxrpl/tx/invariants/MPTInvariant.cpp [hunk 11]
```diff
@@ -231,6 +231,13 @@ ValidMPTIssuance::finalize(
 
         if (hasPrivilege(tx, DestroyMptIssuance))
         {
+            // A VaultDelete that is still cleaning up credentials pinned to its
+            // pseudo-account returns tecINCOMPLETE and has not yet reached the
+            // share issuance. Don't require the issuance to be removed until
+            // the deletion completes (a later transaction).
+            if (rules.enabled(fixCleanup3_4_0) && result == tecINCOMPLETE)
+                return mptIssuancesDeleted_ == 0 && mptIssuancesCreated_ == 0;
```
[the remark this answers is not in the document]
Comment (tyalymov) [reply src/libxrpl/tx/invariants/MPTInvariant.cpp]: Fixed

File: src/libxrpl/tx/Transactor.cpp [hunk 12]
```diff
@@ -1445,6 +1446,11 @@ Transactor::processPersistentChanges(TER result, XRPAmount fee)
         else if (ter == tecINCOMPLETE)
         {
             types.insert(ltRIPPLE_STATE);
+            // A bounded pseudo-account credential cleanup (VaultDelete /
+            // LoanBrokerDelete) persists its partial credential deletions so a
+            // later transaction can resume.
+            if (credentialCleanup)
+                types.insert(ltCREDENTIAL);
```
[the remark this answers is not in the document]
Comment (tyalymov) [reply src/libxrpl/tx/Transactor.cpp:1454]: Fixed

**Review by Tapanito — APPROVED** [review]

**Review by Tapanito — APPROVED** [review]

**Comment by bthomee** [comment]
@mvadari @a1q123456 This still needs one of your approvals before it can be merged.

**Comment by tyalymov** [comment]
@a1q123456 
@mvadari 

Could you please re-review after conflict resolution?

**Review by a1q123456 — APPROVED** [review]
lgtm
